scipy/integrate/_cubature.py

```diff
@@ -0,0 +1,617 @@
+import math
+import heapq
+import itertools
+import functools
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+
+import numpy as np
+
+from numpy.polynomial.polynomial import Polynomial
+
+
+def cub(f, a, b, rule, rtol=1e-05, atol=1e-08, limit=10000, args=()):
+    if limit is None:
+        limit = np.inf
+
+    est = rule.estimate(f, a, b, args)
+    err = rule.error_estimate(f, a, b, args)
+
+    regions = [CubatureRegion(est, err, a, b)]
+    subdivisions = 0
+    success = False
+
+    while np.any(err > atol + rtol * np.abs(est)):
+        region = heapq.heappop(regions)
+
+        est_k = region.estimate
+        err_k = region.error
+
+        a_k, b_k = region.a, region.b
+        m_k = (a_k + b_k) / 2
+
+        subregion_coordinates = zip(
+            _cartesian_product(np.array([a_k, m_k]).T).T,
+            _cartesian_product(np.array([m_k, b_k]).T).T
```

Comment:
if `_cartesian_product` takes a list, can probably simplify here? For one `array_1d.T`  is a noop